lib/Conversion/SeqToSV/FirRegLowering.h

```diff
@@ -16,20 +16,313 @@
 #include "circt/Support/LLVM.h"
 #include "circt/Support/Namespace.h"
 #include "circt/Support/SymCache.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/GraphTraits.h"
+#include <variant>
 
 namespace circt {
+namespace hw {
+namespace detail {
+
+struct SCCNode {
+  Operation *op;
+  llvm::function_ref<bool(Operation *)> filter;
+  unsigned index;
+  explicit SCCNode(Operation *o, llvm::function_ref<bool(Operation *)> f,
+                   unsigned index)
+      : op(o), filter(f), index(index) {}
+  bool operator==(const SCCNode &other) const {
+    return other.op == op && other.index == index;
+  }
+  bool operator!=(const SCCNode &other) const { return !(*this == other); }
+};
+
+// Port iterator is not being used currently, for the purpose of FirRegLowering,
+// only the FirReg are considered as roots for traversal.
+class PortIterator
+    : public llvm::iterator_facade_base<PortIterator, std::forward_iterator_tag,
+                                        SCCNode> {
+  HWModuleOp module;
+  unsigned numPorts;
+  SCCNode sccOp;
+
+public:
+  explicit PortIterator(SCCNode node, bool end = false)
+      : module(cast<HWModuleOp>(node.op)), numPorts(module.getNumPorts()),
+        sccOp(node.op, node.filter, end ? numPorts : 0) {}
+
+  bool operator==(const PortIterator &other) const {
+    return other.sccOp == this->sccOp;
+  }
+  bool operator!=(const PortIterator &other) const { return !(*this == other); }
+  SCCNode operator*() { return sccOp; }
+  PortIterator &operator++() {
+    assert(sccOp.index < numPorts);
+    ++sccOp.index;
+    return *this;
+  }
+  PortIterator operator++(int) {
+    PortIterator result(*this);
+    ++(*this);
+    return result;
+  }
+};
+
+class ModOpIterator
+    : public llvm::iterator_facade_base<ModOpIterator,
+                                        std::forward_iterator_tag, SCCNode> {
+  SCCNode sccOp;
+  HWModuleOp module;
+  Region::OpIterator modOpsIterator, modOpsEnd;
+
+public:
+  explicit ModOpIterator(SCCNode op, bool end = false)
+      : sccOp(op), module(cast<HWModuleOp>(sccOp.op)),
+        modOpsIterator(module.getOps<seq::FirRegOp>().begin()),
```

Comment:
`module.getOps<seq::FirRegOp>()` could walk almost entire IR in the worst case so we should not call it when `end` is true.

Reply:
Thanks a lot Hideto for catching this, this was the reason for the exponential behavior. Fixed it and now the problematic test case passes the SCC phase but fails later on the recursive walk.